### PR TITLE
First review of social token contracts

### PR DESCRIPTION
**Pull request by regexyl**
Here's the first commit of the social tokens contracts and tests. A description of this project can be found on this [Google Docs](https://docs.google.com/document/d/1AcN1t_zBS_YuC_4f_b9B4m4NJ1lhGmWfnXrj0aPYeCo/edit?usp=sharing).

### CODE REVIEW

File: contracts/SocialController.sol [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+    /**
+     * @notice For users to purchase specific creator tokens by quantity. Tokens are minted by each purchase.
+     * @param _tokenId              The token ID of the creator that the user wishes to buy from.
+     * @param _amount               The amount that the user wishes to buy.
+     */
+    function buyToken(uint256 _tokenId, uint256 _amount) external {
+        require(_amount > 0, "You have not input a valid amount for purchase.");
+
+        Creator memory creator = creatorById[_tokenId];
+        uint256 MAX_SAFE_INTEGER = 9007199254740991;
```
Comment (andy-at-mask) [inline contracts/SocialController.sol:172]: - I think the variable name `MAX_SAFE_INTEGER` is kind of confusing.
- I guess, hex(`9007199254740991` ) = '0x1fffffffffffff' is more readable?

File: contracts/PriceCalculator.sol [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+            argument := add(argument, 1)
+            let m := mload(0x40)
+            mstore(m, 0xf8f9cbfae6cc78fbefe7cdc3a1793dfcf4f0e8bbd8cec470b6a28a7a5a3e1efd)
+            mstore(add(m, 0x20), 0xf5ecf1b3e9debc68e1d9cfabc5997135bfb7a7a3938b7b606b5b4b3f2f1f0ffe)
+            mstore(add(m, 0x40), 0xf6e4ed9ff2d6b458eadcdf97bd91692de2d4da8fd2d0ac50c6ae9a8272523616)
+            mstore(add(m, 0x60), 0xc8c0b887b0a8a4489c948c7f847c6125746c645c544c444038302820181008ff)
+            mstore(add(m, 0x80), 0xf7cae577eec2a03cf3bad76fb589591debb2dd67e0aa9834bea6925f6a4a2e0e)
+            mstore(add(m, 0xa0), 0xe39ed557db96902cd38ed14fad815115c786af479b7e83247363534337271707)
+            mstore(add(m, 0xc0), 0xc976c13bb96e881cb166a933a55e490d9d56952b8d4e801485467d2362422606)
+            mstore(add(m, 0xe0), 0x753a6d1b65325d0c552a4d1345224105391a310b29122104190a110309020100)
```
Comment (andy-at-mask) [inline contracts/PriceCalculator.sol:151]: I did a quick test, it looks like this one is pretty efficient(in terms of gas consumption).
But I need a little more time to digest what the code is really doing here[we need to understand it thoroughly].
As someone else was asking [`please explain!`](https://ethereum.stackexchange.com/questions/8086/logarithm-math-operation-in-solidity).

File: contracts/SocialController.sol [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+                constantFactor,
+                numOfOutstandingTokens,
+                _amount
+            );
+        }
+
+        _burn(msg.sender, _tokenId, _amount);
+        creatorById[_tokenId].numOfOutstandingTokens -= _amount;
+
+        IERC20 dai = IERC20(0x6B175474E89094C44Da98b954EedeAC495271d0F);
```
Comment (andy-at-mask) [inline contracts/SocialController.sol:241]: Here and `LINE199` have duplicated `magic numbers`. Using `a constant variable` might be better.

File: test/SocialController.ts [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    this.signers.miscellaneous = signers[7];
+
+    // Get DAI interface
+    this.daiMainnetContract = <IDai>await ethers.getContractAt("IDai", DAI_MAINNET_ADDRESS);
+
+    // For impersonating DAI_HOLDER_ADDRESS_1
+    await hre.network.provider.request({
+      method: "hardhat_impersonateAccount",
+      params: [DAI_HOLDER_ADDRESS_1],
+    });
```
Comment (andy-at-mask) [inline test/SocialController.ts:60]: I think your approach to testing 👍

File: contracts/PriceCalculator.sol [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+        uint256 _growthFactor,
+        uint256 _constantFactor,
+        uint256 _numOfOutstandingTokens,
+        uint256 _amountToSell
+    ) internal pure returns (uint256) {
+        uint256 totalConstant = _amountToSell * _constantFactor;
+        uint256 totalNonConstant;
+
+        unchecked {
+            for (uint256 i = 0; i < _amountToSell; i++) {
```
Comment (andy-at-mask) [inline contracts/PriceCalculator.sol:91]: Is it possible to get rid of this loop[call: `_subFormulaPower` only once]. 
I have not tried yet, but according to your `bonding curve formula`, I guess it is worth a try.
If it is, then we can reduce gas consumption.

File: contracts/SocialController.sol (hunk 1 above)
Comment (regexyl) [reply contracts/SocialController.sol:172]: Updated var name to JS_SAFE_INT and shifted it to become a constant state variable.

File: contracts/SocialController.sol (hunk 3 above)
Comment (regexyl) [reply contracts/SocialController.sol:241]: Updated to constant

**Pull request closed without merging**